actors/evm/src/interpreter/instructions/lifecycle.rs

```diff
@@ -161,3 +161,418 @@ pub fn selfdestruct(
+                    ExitCode::OK,
+                    Some(ErrorNumber::IllegalOperation),
+                );
+            }
+            (m) {
+                // input data
+                PUSH4; 0x01; 0x02; 0x03; 0x04;
+                PUSH0;
+                MSTORE;
+                // the deed
+                CREATE;
+            }
+            m.state.stack.push(U256::from(4)).unwrap();    // input size
+            m.state.stack.push(U256::from(28)).unwrap();   // input offset
+            m.state.stack.push(U256::from(1234)).unwrap(); // initial value
+            for _ in 0..4 {
+                m.step().expect("execution step failed");
+            }
+            assert_eq!(m.state.stack.len(), 1);
+            assert_eq!(m.state.stack.pop().unwrap(), U256::from(0));
+            assert_eq!(m.system.nonce, 2);
+        };
+    }
+
+    #[test]
+    fn test_selfdestruct() {
+        // tests the outcome of selfdestruct
+        let beneficiary = EthAddress::from_id(1001);
+        let fil_beneficiary = FilAddress::new_id(1001);
+
+        evm_unit_test! {
+            (rt) {
+                rt.set_balance(TokenAmount::from_atto(1_000_000));
+
+                rt.expect_send(
+                    fil_beneficiary,
+                    METHOD_SEND,
+                    None,
+                    TokenAmount::from_atto(1_000_000),
+                    None,
+                    SendFlags::empty(),
+                    None,
+                    ExitCode::OK,
+                    None,
+                );
+            }
+            (m) {
+                SELFDESTRUCT;
+                // correctness check sled
+                INVALID;
+            }
+            m.state.stack.push(beneficiary.as_evm_word()).unwrap();
+            let result = m.execute().expect("execution failed");
+            assert_eq!(result.outcome, crate::Outcome::Return);
+            assert_eq!(result.return_data.len(), 0);
+        }
+    }
+
+    #[test]
+    fn test_selfdestruct_tombstone() {
+        // tests the selfdestruct step to ensure a tombstone is created.
+        // can't merge these two tests because rust....
+        let beneficiary = EthAddress::from_id(1001);
+        let fil_beneficiary = FilAddress::new_id(1001);
+
+        evm_unit_test! {
+            (rt) {
+                rt.set_balance(TokenAmount::from_atto(1_000_000));
+
+                rt.expect_send(
+                    fil_beneficiary,
+                    METHOD_SEND,
+                    None,
+                    TokenAmount::from_atto(1_000_000),
+                    None,
+                    SendFlags::empty(),
+                    None,
+                    ExitCode::OK,
+                    None,
+                );
+            }
+            (m) {
+                SELFDESTRUCT;
+            }
+            m.state.stack.push(beneficiary.as_evm_word()).unwrap();
+            m.step().expect("execution step failed");
+            assert!(m.system.tombstone.is_some());
```

Comment:
Let's check the epoch/origin.

Reply:
done; mock runtime always returns 0 for nonce, but I did set the origin and it stuck around as it should.